File: 112-path-sum/path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        """ T: O(n)O(n)O(n) since we have to look at every node it the binary tree
            S: O(h) height of tree, worst case O(n), if tree is balanced O(logn)
        """

        # inorder DFS
        def dfs(node, curSum):
            # empty tree base case
            if not node:
                return False
            
            curSum += node.val
            # check if does NOT have leaves and if current sum is the target
            if not node.left and not node.right:
                return curSum == targetSum

            return (dfs(node.left, curSum) or 
                    dfs(node.right, curSum))

        # pass in root node and current sum of 0
        return dfs(root, 0)

File: 112-path-sum/path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        return dfs(root, targetSum, 0);
    }

    bool dfs(TreeNode* node, int target, int curr){
        if(!node) return false;
        curr += node->val;
        if(!node->left && !node->right) return target == curr;
        return (dfs(node->left, target, curr) || dfs(node->right, target, curr));
    }
};